### ISSUE TITLE
represent dependencies on previous sibling elements for Transformer

### ISSUE THREAD

**Issue by dylemma** [dylemma/xml-spac#7]
I encountered a situation like

```scala
case class Info(id: String, value: String)
case class Data(info: Info, otherStuff: String)
```
```xml
<asdf>
  <infos>
    <info id="a">abc</info>
    ...
    <info id="z">zyx</info>
  </infos>
  <datas>
    <data info_id="x">...</data>
    <!-- lots and lots of datas -->
    <data info_id="b">...</data>
  </datas>
</asdf>
```
and I wanted to create a `Transformer[XMLEvent, Data]` (as opposed to a `Parser[XMLEvent, List[Data]]`, since there were enough datas that I'd prefer to avoid having them all in memory at once.

The current capabilities of xml-spac make this impractical. I ultimately worked around the problem by creating a mutable map in which to store the Infos, then referencing that map from the Data transformer, but that feels dirty (a huge point of this library is to prevent client code from needing to manage mutable state).

XML-SPaC needs a way to express the Data transformer's dependency on its preceding sibling(s) without relying on the client to manage the internal state of that dependency. The best syntax for this that I've been able to think up so far has been:

```scala
val datasTransformer = for {
  infosMap <- infosMapParser.inSequence
} yield makeDatasTransformer(infosMap)
```

where `Parser#inSequence` returns some `flatMap/map`-able object i.e.

```scala
class ParserInSequence[Ctx, Out](parser: Parser[Ctx, Out]){
  def map[B](f: Out => Transformer[XMLEvent, B]): Transformer[XMLEvent, B]
  def flatMap[B]( /* same signature as map */)
}
```

Using this approach it should be possible to represent large chains, e.g.

```scala
for {
  results1 <- parser1.inSequence
  ...
  resultsN <- makeParserN(<previous results>).inSequence
} yield makeTransformer(results1 .. resultsN)
```

This does seem very similar to `Parser.combine`, but the key difference is that it results in a `Transformer` instead of a `Parser`.

**Comment by dylemma** [comment]
I had an idea for representing the "sequence" stream.

The difficulty was in that each parser typically has its own "scope" given to it by a Splitter, but that scope is generally not exposed at the type level; this makes it impossible to say something like "only feed events to this parser when its specific context is active".

The idea to get around this is to hold onto "scoping" events (i.e. StartElements) so that they can be replayed when the 1st parser claims to be complete and the 2nd parser is ready to start.

So in the example XML

```xml
<parent>
  <child1>Hello</child1>
  <child2>Goodbye</child2>
</parent>
```

with 2 parsers (one for the text in child1, the second for the text in child2), the following could happen:

 - `StartElement("parent")` passed to Parser1
 - `StartElement("child1")` passed to Parser1
 - `Characters("Hello")` passed to Parser1
 - `EndElement("child1")` passed to Parser1
 - Parser1 yields a result, so Parser2 can become active, and can receive Parser1's result
 - `StartElement("parent")` passed via replay to Parser2
 - (end of replay stack, continue with the normal event stream)
 - `StartElement("child2")` passed to Parser2
 - `Characters("Goodbye")` passed to Parser2
 - `EndElement("child2") passed to Parser2
 - Parser2 yields a result, completing the combined parser that uses Parser1+Parser2
 - `EndElement("parent")` would be ignored

**Comment by dylemma** [comment]
This is implemented, and will be in release 0.4

**Comment by dylemma** [comment]
Done, see https://github.com/dylemma/xml-spac/blob/master/examples/src/main/scala/io/dylemma/xml/example/Example5_FollowedBy.scala for an example
